### ISSUE TITLE
The result of format_annotation() has been changed with python 3.7.0a4+

### ISSUE THREAD

**Issue by tk0miya** [sphinx-doc/sphinx#4490]
### Problem
- The result of format_annotation() has been changed with python 3.7.0a4+.

#### Procedure to reproduce the problem
- Run tests of Sphinx with python 3.7.0a4+

#### Error logs / results
https://travis-ci.org/sphinx-doc/sphinx/jobs/332762559#L2003

**Comment by tk0miya** [comment]
Since py370a4+, the implementation of `typing` module has been changed (ref: https://github.com/python/cpython/pull/4906)
```
$ python3.6
Python 3.6.4 (default, Dec 20 2017, 01:09:24)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> isinstance(typing.List, type)
True
>>> typing.GenericMeta
<class 'typing.GenericMeta'>
>>> typing.CallableMeta
<class 'typing.CallableMeta'>
```
```
$ python3.7
Python 3.7.0a4+ (heads/master:131fd7f, Jan 24 2018, 23:03:18)
[Clang 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> isinstance(typing.List, type)
False
>>> typing.GenericMeta
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typing' has no attribute 'GenericMeta'
>>> typing.CallableMeta
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typing' has no attribute 'CallableMeta'
```

For this reason, the result of `format_annotation()` has been changed.
```
>       assert formatargspec(f, *getargspec(f)) == expected
E       AssertionError: assert '(x: typing.L...st[-T_contra]' == '(x: typing.Li...ist[T_contra]'
E         - (x: typing.List[~T], y: typing.List[+T_co], z: T) -> typing.List[-T_contra]
E         ?                 -                   -                            -
E         + (x: typing.List[T], y: typing.List[T_co], z: T) -> typing.List[T_contra]
```
There are `+`, `-` and `~` prefix on annotation.

**Comment by tk0miya** [comment]
Note: I just pushed workarounds for this;  #4488 and #4489 to pass test cases temporarily.

**Comment by tk0miya** [comment]
Fixed by #4502
